src/ripple/protocol/impl/SecretKey.cpp

```diff
@@ -54,45 +53,160 @@ SecretKey::to_string() const
+        std::copy(generator_.begin(), generator_.end(), buf.begin());
+        copy_uint32(buf.data() + 33, seq);
+
+        // The odds that this loop executes more than once are neglible
+        // but we impose a maximum limit just in case.
+        for (std::uint32_t subseq = 0; subseq != 128; ++subseq)
+        {
+            copy_uint32(buf.data() + 37, subseq);
+
+            auto const ret = sha512Half_s(buf);
+
+            if (secp256k1_ec_seckey_verify(secp256k1Context(), ret.data()) == 1)
+            {
+                secure_erase(buf.data(), buf.size());
+                return ret;
+            }
+        }
+
+        Throw<std::runtime_error>("Unable to derive generator from seed");
+    }
 
 public:
     explicit Generator(Seed const& seed)
+        : root_(deriveDeterministicRootKey(seed))
     {
-        // FIXME: Avoid copying the seed into a uint128 key only to have
-        //        generateRootDeterministicPublicKey copy out of it.
-        uint128 ui;
-        std::memcpy(ui.data(), seed.data(), seed.size());
-        gen_ = generateRootDeterministicPublicKey(ui);
+        secp256k1_pubkey pubkey;
+        if (secp256k1_ec_pubkey_create(
+                secp256k1Context(), &pubkey, root_.data()) != 1)
+            LogicError("derivePublicKey: secp256k1_ec_pubkey_create failed");
+
+        auto len = generator_.size();
+
+        if (secp256k1_ec_pubkey_serialize(
+                secp256k1Context(),
+                generator_.data(),
+                &len,
+                &pubkey,
+                SECP256K1_EC_COMPRESSED) != 1)
+            LogicError("derivePublicKey: secp256k1_ec_pubkey_serialize failed");
     }
 
-    /** Generate the nth key pair.
+    ~Generator()
+    {
+        secure_erase(root_.data(), root_.size());
+        secure_erase(generator_.data(), generator_.size());
+    }
 
-        The seed is required to produce the private key.
-    */
+    /** Generate the nth key pair. */
     std::pair<PublicKey, SecretKey>
-    operator()(Seed const& seed, std::size_t ordinal) const
+    operator()(std::size_t ordinal) const
     {
-        // FIXME: Avoid copying the seed into a uint128 key only to have
-        //        generatePrivateDeterministicKey copy out of it.
-        uint128 ui;
-        std::memcpy(ui.data(), seed.data(), seed.size());
-        auto gsk = generatePrivateDeterministicKey(gen_, ui, ordinal);
-        auto gpk = generatePublicDeterministicKey(gen_, ordinal);
-        SecretKey const sk(Slice{gsk.data(), gsk.size()});
-        PublicKey const pk(Slice{gpk.data(), gpk.size()});
-        secure_erase(ui.data(), ui.size());
-        secure_erase(gsk.data(), gsk.size());
-        return {pk, sk};
+        // Generates Nth secret key:
+        auto gsk = [this, tweak = calculateTweak(ordinal)]() {
+            auto rpk = root_;
+
+            if (secp256k1_ec_privkey_tweak_add(
+                    secp256k1Context(), rpk.data(), tweak.data()) == 1)
```

Comment:
The old code did not check for zero, and this new code does. Of course, this is an *extremely* unlikely case to hit. I'm just pointing out a functional difference between the old code and the new.

Reply:
You're right: this is a difference between the old and new code. Sharp eyes. The odds of this happening (as you point out) are, practically, 0 and I don't think it's particularly helpful to clutter the code up checking for this, but if the consensus is that this is important enough, I'm happy to make the change.